File: sdks/official/fraiseql-php/conformance/export.php
<?php

/**
 * Author the cross-SDK conformance fixture with the PHP SDK's public API.
 *
 * Driven by sdks/official/conformance/run.py; see sdks/official/conformance/README.md.
 *
 * The one rule for every SDK's copy of this file: author through the SDK, never
 * hand-assemble the JSON. This goes through the attribute-based idiom the
 * SchemaExporter docblock documents — StaticAPI::register() plus the query/mutation
 * builders — and exports with SchemaExporter, which is what `bin/fraiseql export` runs.
 */

declare(strict_types=1);

require_once __DIR__ . '/../vendor/autoload.php';

use FraiseQL\ActorType;
use FraiseQL\Attributes\GraphQLField;
use FraiseQL\Attributes\GraphQLType;
use FraiseQL\SchemaExporter;
use FraiseQL\StaticAPI;
use FraiseQL\VectorConfig;

#[GraphQLType(name: 'User', sqlSource: 'v_user', relay: true)]
final class ConformanceUser
{
    #[GraphQLField(type: 'ID', nullable: false)]
    public string $id;

    #[GraphQLField(type: 'String', nullable: false)]
    public string $email;

    #[GraphQLField(
        type: 'String',
        nullable: true,
        description: 'The user\'s "display" name',
        deprecated: 'use displayName',
    )]
    public ?string $name;

    #[GraphQLField(type: 'Float', nullable: true, scope: 'read:User.salary')]
    public ?float $salary;

    // Two words and a digit segment (#1249). A PHP property is idiomatically
    // camelCase and is emitted verbatim, so these match the reference as written;
    // the SDKs that translate are the ones with snake_case or PascalCase
    // identifiers (Python, Ruby, Elixir, C#, F#).
    #[GraphQLField(type: 'String', nullable: true)]
    public ?string $lastLoginAt;

    #[GraphQLField(type: 'String', nullable: true)]
    public ?string $phone1;
}

#[GraphQLType(name: 'Order', sqlSource: 'v_order')]
final class ConformanceOrder
{
    #[GraphQLField(type: 'ID', nullable: false)]
    public string $id;

    #[GraphQLField(type: 'Float', nullable: false)]
    public float $total;

    #[GraphQLField(type: 'String', nullable: false)]
    public string $status;
}

// `crud` is an authoring-time expansion the compiler has no concept of, so the only
// evidence this SDK implements it is that the operations and input objects appear in the
// compiled schema. `computed` is the same: emitting the flag makes the document
// uncompilable, so the sole evidence it was honoured is `slug` on the type and
// absent from both input objects.
#[GraphQLType(name: 'SupportTicket', sqlSource: 'v_support_ticket', crud: true)]
final class ConformanceSupportTicket
{
    #[GraphQLField(type: 'Int', nullable: false)]
    public int $id;

    #[GraphQLField(type: 'String', nullable: false)]
    public string $title;

    #[GraphQLField(type: 'String', nullable: false)]
    public string $dueDate;

    #[GraphQLField(type: 'String', nullable: false, computed: true)]
    public string $slug;
}

#[GraphQLType(name: 'UserNotFound', sqlSource: 'v_user_not_found', isError: true)]
final class ConformanceUserNotFound
{
    #[GraphQLField(type: 'String', nullable: false)]
    public string $message;

    #[GraphQLField(type: 'String', nullable: false)]
    public string $code;
}

#[GraphQLType(name: 'Document', sqlSource: 'v_document')]
final class ConformanceDocument
{
    #[GraphQLField(type: 'ID', nullable: false)]
    public string $id;

    #[GraphQLField(type: 'Vector', nullable: false, vectorConfig: new VectorConfig(
        dimensions: 1536,
        indexType: VectorConfig::INDEX_IVF_FLAT,
        distanceMetric: VectorConfig::METRIC_L2,
    ))]
    public array $embedding;

    #[GraphQLField(type: 'BitVector', nullable: false, vectorConfig: new VectorConfig(
        dimensions: 768,
        distanceMetric: VectorConfig::METRIC_HAMMING,
    ))]
    public string $fingerprint;

    #[GraphQLField(type: 'HalfVector', nullable: true, vectorConfig: new VectorConfig(
        dimensions: 1536,
        distanceMetric: VectorConfig::METRIC_INNER_PRODUCT,
    ))]
    public ?array $compact;

    #[GraphQLField(type: 'SparseVector', nullable: true, vectorConfig: new VectorConfig(
        dimensions: 30000,
        indexType: VectorConfig::INDEX_NONE,
    ))]
    public ?string $terms;

    #[GraphQLField(type: 'Float', nullable: false, vectorDistance: 'embedding')]
    public float $similarity;
}

#[GraphQLType(name: 'CreateUserInput', isInput: true)]
final class ConformanceCreateUserInput
{
    #[GraphQLField(type: 'String', nullable: false)]
    public string $email;

    #[GraphQLField(type: 'String', nullable: true)]
    public ?string $name;
}

#[GraphQLType(name: 'User', sqlSource: 'v_user')]
final class ConformanceMinimalUser
{
    #[GraphQLField(type: 'ID', nullable: false)]
    public string $id;

    #[GraphQLField(type: 'String', nullable: false)]
    public string $email;
}

function authorMinimal(): void
{
    StaticAPI::register(ConformanceMinimalUser::class);

    StaticAPI::query('users')
        ->returnType('User')
        ->returnsList()
        ->nullable(false)
        ->sqlSource('v_user')
        ->register();
}

function authorFull(): void
{
    StaticAPI::register(ConformanceUser::class);
    StaticAPI::register(ConformanceOrder::class);
    StaticAPI::register(ConformanceSupportTicket::class);
    StaticAPI::register(ConformanceUserNotFound::class);
    StaticAPI::register(ConformanceDocument::class);
    StaticAPI::register(ConformanceCreateUserInput::class);

    StaticAPI::enum('OrderStatus', ['PENDING', 'SHIPPED', 'CANCELLED']);

    StaticAPI::query('users')
        ->returnType('User')
        ->returnsList()
        ->nullable(false)
        ->sqlSource('v_user')
        ->register();

    StaticAPI::query('user')
        ->returnType('User')
        ->returnsList(false)
        ->nullable(true)
        ->sqlSource('v_user')
        ->argument('id', 'ID', nullable: false)
        ->register();

    StaticAPI::query('tenantOrders')
        ->returnType('Order')
        ->returnsList()
        ->nullable(false)
        ->sqlSource('v_order')
        ->inject(['tenant_id' => 'jwt:tenant_id'])
        ->cacheTtlSeconds(300)
        ->requiresRole('admin')
        // #966's actor allow-list, enforced in the same executor gate as requiresRole on
        // every transport, and authorable in no SDK until #1123.
        ->requiresActor([ActorType::HUMAN_USER, ActorType::SERVICE_ACCOUNT])
        ->register();

    StaticAPI::mutation('createUser')
        ->returnType('User')
        ->sqlSource('fn_create_user')
        ->operation('insert')
        ->argument('email', 'String', nullable: false)
        ->argument('name', 'String', nullable: true)
        ->invalidatesViews(['v_user', 'v_user_summary'])
        ->invalidatesFactTables(['tf_signup'])
        ->requiresActor([ActorType::SERVICE_ACCOUNT])
        ->register();

    StaticAPI::mutation('placeOrder')
        ->returnType('Order')
        ->sqlSource('fn_place_order')
        ->operation('insert')
        ->inject(['user_id' => 'jwt:sub'])
        ->invalidatesViews(['v_order_summary'])
        ->invalidatesFactTables(['tf_sale'])
        ->register();

    StaticAPI::subscription('orderUpdated')
        ->entityType('Order')
        ->argument('orderId', 'ID', argNullable: true)
        ->description('Stream of order update events')
        ->topic('order_events')
        ->filterCondition('orderId', '$.id')
        ->fields(['id', 'total'])
        ->build();
}

$fixture = getenv('FRAISEQL_CONFORMANCE_FIXTURE');
$out     = getenv('FRAISEQL_CONFORMANCE_OUT');
if ($fixture === false || $out === false) {
    fwrite(STDERR, "FRAISEQL_CONFORMANCE_FIXTURE and FRAISEQL_CONFORMANCE_OUT must be set\n");
    exit(2);
}

StaticAPI::clear();
match ($fixture) {
    'minimal' => authorMinimal(),
    'full'    => authorFull(),
    default   => throw new RuntimeException("unknown fixture $fixture"),
};

SchemaExporter::exportToFile($out);

File: sdks/official/fraiseql-elixir/conformance/export.exs
# Authors the cross-SDK conformance fixture with the Elixir SDK's public API.
#
# Driven by `sdks/official/conformance/run.py`; see
# `sdks/official/conformance/README.md`.
#
# The one rule for every SDK's copy of this file: author through the SDK, never
# hand-assemble the JSON.
#
# Note the naming idiom: `fraiseql_type` takes a string, while `fraiseql_query` and
# `fraiseql_mutation` take atoms — the query name is the atom verbatim and the mutation
# name is its camelCase form, so `:place_order` declares `placeOrder`. The pre-existing `generate_parity_schema_test.exs` built the
# expected document as a literal map and never touched `FraiseQL.SchemaExporter`, so it
# was structurally incapable of failing.

defmodule Conformance.MinimalSchema do
  use FraiseQL.Schema

  fraiseql_type "User", sql_source: "v_user" do
    field(:id, :id, nullable: false)
    field(:email, :string, nullable: false)
  end

  fraiseql_query(:users,
    return_type: "User",
    returns_list: true,
    nullable: false,
    sql_source: "v_user"
  )
end

defmodule Conformance.FullSchema do
  use FraiseQL.Schema

  fraiseql_type "User", sql_source: "v_user", relay: true do
    field(:id, :id, nullable: false)
    field(:email, :string, nullable: false)
    field(:name, :string,
      nullable: true,
      description: ~s(The user's "display" name),
      deprecated: "use displayName"
    )
    field(:salary, :float, nullable: true, requires_scope: "read:User.salary")
  end

  fraiseql_type "Order", sql_source: "v_order" do
    field(:id, :id, nullable: false)
    field(:total, :float, nullable: false)
    field(:status, :string, nullable: false)
  end

  # `crud` is an authoring-time expansion the compiler has no concept of, so the only
  # evidence this SDK implements it is that the operations and input objects appear in the
  # compiled schema. `computed` is the same: emitting the flag makes the document
  # uncompilable, so the sole evidence it was honoured is `slug` on the type and absent
  # from both input objects.
  fraiseql_type "SupportTicket", sql_source: "v_support_ticket", crud: true do
    field :id, :int, nullable: false
    field :title, :string, nullable: false
    field :slug, :string, nullable: false, computed: true
  end

  fraiseql_type "UserNotFound", sql_source: "v_user_not_found", is_error: true do
    field(:message, :string, nullable: false)
    field(:code, :string, nullable: false)
  end

  fraiseql_type "Document", sql_source: "v_document" do
    field(:id, :id, nullable: false)

    field(:embedding, :vector,
      nullable: false,
      vector_config: [dimensions: 1536, index_type: :ivf_flat, distance_metric: :l2]
    )

    field(:fingerprint, :bit_vector,
      nullable: false,
      vector_config: [dimensions: 768, distance_metric: :hamming]
    )

    field(:compact, :half_vector,
      nullable: true,
      vector_config: [dimensions: 1536, distance_metric: :inner_product]
    )

    field(:terms, :sparse_vector,
      nullable: true,
      vector_config: [dimensions: 30000, index_type: :none]
    )

    field(:similarity, :float, nullable: false, vector_distance: :embedding)
  end

  # `is_input: true` is this SDK's only route to an input object — its exporter emits no
  # `input_types` key, and the compiler reclassifies a type carrying the flag.
  fraiseql_type "CreateUserInput", is_input: true do
    field(:email, :string, nullable: false)
    field(:name, :string, nullable: true)
  end

  fraiseql_enum("OrderStatus", values: ["PENDING", "SHIPPED", "CANCELLED"])

  fraiseql_query(:users,
    return_type: "User",
    returns_list: true,
    nullable: false,
    sql_source: "v_user"
  )

  fraiseql_query :user,
                 return_type: "User",
                 returns_list: false,
                 nullable: true,
                 sql_source: "v_user" do
    argument(:id, :id, nullable: false)
  end

  fraiseql_query(:tenantOrders,
    return_type: "Order",
    returns_list: true,
    nullable: false,
    sql_source: "v_order",
    inject_params: %{"tenant_id" => "jwt:tenant_id"},
    cache_ttl_seconds: 300,
    requires_role: "admin",
    # #966's actor allow-list, enforced in the same executor gate as requires_role on every
    # transport, and authorable in no SDK until #1123.
    requires_actor: ["human_user", "service_account"]
  )

  fraiseql_mutation :create_user,
                    return_type: "User",
                    sql_source: "fn_create_user",
                    operation: "insert",
                    invalidates_views: ["v_user", "v_user_summary"],
                    invalidates_fact_tables: ["tf_signup"],
                    requires_actor: ["service_account"] do
    argument(:email, :string, nullable: false)
    argument(:name, :string, nullable: true)
  end

  fraiseql_mutation(:place_order,
    return_type: "Order",
    sql_source: "fn_place_order",
    operation: "insert",
    inject_params: %{"user_id" => "jwt:sub"},
    invalidates_views: ["v_order_summary"],
    invalidates_fact_tables: ["tf_sale"]
  )
end

fixture = System.get_env("FRAISEQL_CONFORMANCE_FIXTURE")
out = System.get_env("FRAISEQL_CONFORMANCE_OUT")

if is_nil(fixture) or is_nil(out) do
  IO.puts(:stderr, "FRAISEQL_CONFORMANCE_FIXTURE and FRAISEQL_CONFORMANCE_OUT must be set")
  System.halt(2)
end

module =
  case fixture do
    "minimal" -> Conformance.MinimalSchema
    "full" -> Conformance.FullSchema
    other -> raise ArgumentError, "unknown fixture #{other}"
  end

:ok = FraiseQL.SchemaExporter.export_to_file!(module, out)
